engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageFunctionGenerator.java

```diff
@@ -679,118 +362,666 @@ public static void main(String args[]) {
+                break;
+            case GREATER_EQUALS:
+                sb.append("        return compareTo(a, b) >= 0;\n");
+                break;
+            default:
+                throw new IllegalStateException("Unsupported operator: " + op);
+        }
+
+        sb.append("    }\n\n");
+
+        return sb.toString();
     }
 
-    private static void appendTest(StringBuilder buf, MessageFormat messageFormat, BinaryExpr.Operator op,
-            Class<?> type1, Class<?> type2) {
-        append(buf, messageFormat, op, null, type1, type2, getLiteral(type1), getLiteral(type2));
+    private static String generateCastFunction(@NotNull final Class<?> classA, @NotNull final Class<?> classB) {
+        final StringBuilder sb = new StringBuilder();
+
+        sb.append(MessageFormat.format("    public static {1} {1}Cast({0} a) '{'\n",
+                classA.getSimpleName(),
+                classB.getSimpleName()));
+
+        final String nullA = nullForType(classA);
+        final String nullB = nullForType(classB);
+
+        sb.append(MessageFormat.format("" +
+                "        return a == {0} ? {1} : ({2})a;\n",
+                nullA, nullB, classB.getSimpleName()));
+        sb.append("    }\n\n");
+
+        return sb.toString();
     }
 
-    private static void appendTest(StringBuilder buf, MessageFormat messageFormat,
-            Class<?> type1, Class<?> type2,
-            String literal1, String literal2) {
-        append(buf, messageFormat, null, null, type1, type2, literal1, literal2);
+    private static String generateCastFromObjFunction(@NotNull final Class<?> classA) {
+        final StringBuilder sb = new StringBuilder();
+
+        final String typeA = classA.getSimpleName();
+        sb.append(MessageFormat.format("    public static {0} {0}Cast(Object a) '{'\n",
+                typeA));
+
+        final String nullA = nullForType(classA);
+        if (classA == char.class) {
+            sb.append(MessageFormat.format("" +
+                    "        return a == null ? {0} : (char)a;\n",
+                    nullA));
+        } else {
+            sb.append(MessageFormat.format("" +
+                    "        return a == null ? {0} : ((Number) a).{1}Value();\n",
+                    nullA, typeA));
+        }
+        sb.append("    }\n\n");
+
+        return sb.toString();
     }
 
-    private static void append(StringBuilder buf, MessageFormat messageFormat, BinaryExpr.Operator op,
-            String opDescription, Class<?> type1, Class<?> type2, String literal1, String literal2) {
-        Class<?> promotedType;
+    private static String generateNegateFunction(@NotNull final Class<?> classA) {
+        final StringBuilder sb = new StringBuilder();
 
+        final Class<?> returnType;
+        if (classA == int.class || classA == char.class || classA == byte.class || classA == short.class) {
+            returnType = int.class;
```

Comment:
seems odd that a short turns into an int.

I understand char getting promoted however.